### ISSUE TITLE
Removing modal windows - letters/emails

### ISSUE THREAD

**Issue by LaurenceLeo** [hezmondo/Jinn#125]
One of the longest running complaints about Jinn is the inability to click away from a letter, perform another task, then return to the letter. 

This is an issue that effects almost all of our users daily. 

Since there is some danger in removing modality I propose some safeguards:

- Only one letter preview screen can be opened at a time
- We do not extend this to pay requests

I will do some investigation into SQL tables that could be effected by bad user behaviour.

If anyone has anything else to add please do.

**Comment by jnbarchan** [comment]
This was done by L., but unfortunately no documentation here as to what, why & how....  Also, it *was* extended to pay requests (though possibly only to single pay request, not multi-pay request run).

The implementation is flawed.  For example, in one route it works as follows:

* From **Write Letter** dialog, clicking **Preview/merge** does a `.show()` on `HtmlPreviewDialog` (i.e. modeless), which is fine.
* However, it also does a `.show()` (modeless) on the invoking `LetterDialog`, and that does a `self.hide()` while the preview window is shown, and when the preview dialog gets closed that does a `self.parent().show()` to re-show the hidden parent.
* This relies on passing the parent dialog to child modeless dialog.
* Without the parent `hide()`, on all platforms that will make the modeless always appear on top of the parent: whether it blocks or not, we cannot have this in the user interface, the preview dialog must not sit above other Jinn windows where the user might want to look.
* With the parent `hide()`, while the parent is hidden it actually works as desired under Linux, in that the modeless child does *not* sit on top of the grandparent (given that the parent is hidden).  However under Windows it now sits on top of the grandparent, which we cannot have.

The correct way to have a top-level modeless dialog (i.e. never sits on top of any other window) is to pass `None` as the parent.  We need to change over to that principle.

The simplest way to do it seems to be to make *only* the `HtmlPreviewDialog` modeless.  Agreed with H. that is all end-users are asking for.

This means that anywhere which currently displays the `HtmlPreviewDialog`  *and then waits for it to exit in order to perform further processing* (e.g. according as which button user pressed to save/exit from preview) cannot do that any longer.  Such code must, in some or shape or form, be passed to the preview dialog so that it gets executed there on button press.

**Comment by LaurenceLeo** [comment]
Hey Jonathan. Let me provide a basic rundown on what I was doing. I had reorganised the letter system so that it no longer relied upon .exec() to hold up the code whilst a dialog was open. This let me experiment with the show() methods. I found that I roughly had the behaviour I desired. I acknowledge that it is far from perfect. At the time I was a bit obsessed with making everything modeless which was a mistake. This is why I ended up making the letterdialog hide and re-show itself. It isn't the best way to do this. 

This resulted in several undesired behaviours. Modeless dialogs always had to be "always on top" of their parents, modeless dialogs were not correctly handled by the system tray/taskbar. The reason I ended up assigning parents to shown dialogs was because in some situations the parents would closes and be garbage collected. This meant that the dialogs would be shown for a second before they closed themselves.

I found a way to work around this by using Comms to show() the dialogs I wanted to be modeless from **Main**, since **Main** is never cleared by Python. This worked much better in my testing. I have since backed off this issue since you are looking into it and I don't want to make things any more confusing.

If you look in **home.py** at lines **688, 853** there is some of what I was planning on using to open modeless dialogs. It might be worth implementing a show() type function into **Comms** so that we can show() dialogs using **Main** when necessary?

**Comment by jnbarchan** [comment]
While doing this, I introduced a means (for development) of *pretending* to send emails/copying to email draft folder while never actually doing so:

In  `settings.ini`, `[Debug]` section:
```
# pretend_send_email: set to 1 to disable actually sending any emails (for development)
pretend_send_email = 0
```
Setting this to `1` allows you to test all the code for sending emails without risking it actually sending them out.  Messages/logs will include the word `*pretend*` so that you know.  You still need to set up SMTP/IMAP settings properly as necessary: the code does still connect to SMTP/IMAP (so you know whether that part is working correctly), but will not actually send/add into drafts folder when it comes to do that.  Obviously, you will not want to run with that when using Jinn "live"!

**Comment by jnbarchan** [comment]
Well, I claim to have done it, as best I can.  The code is a quagmire.  The two big differences are:

* The only dialog which is shown modelessly is `HtmlPreviewDialog`.

* There is no use of "parents" (`self.parent()....`), which was all over the place, any longer.

The code broadly works as before, but there were compromises in a few places.  It involved substantial re-architecting.  I hope I never have to deal with stuff like this again :(

Good luck to whoever tests it.  It needs thorough testing all over the place, for cases which do not work as well as those which do.

**Comment by jnbarchan** [comment]
H. reported: under Windows, when the modeless Preview window gets opened and at the same time the invoking modal dialog gets closed causing return to the main window, the main window is shown *on top* of the modeless.  (You can still click between them OK to bring either up-front, but we want the user to be left with the modeless preview on top initially.)

I did encounter this behaviour initially under Linux.  So in `common/comms.py`, function `showModelessHtmlPreviewDialog()` I already changed to go:
```
# show dialog modelessly, no parent
preview.show()
...
# bring modeless dialog to front, else it goes behind main window
preview.raise_()
preview.activateWindow()
```
The `raise_()` is supposed to bring the newly-created modeless on top explicitly, it's the way to do it and makes it work under Linux.  But seemingly not under Windows....

It might be that this is a timing issue, behaving differently under Windows/faster machine: maybe the `raise_()` is being called too early, immediately after the modeless starts to get shown, and only after that the modal gets closed, bringing it/its main window parent back on top?

**Comment by LaurenceLeo** [comment]
I can confirm that I found the same issue here on Windows 10. I was researching it last night. If it is an issue with Windows it might be the **ForegroundLockTimeout**.

Sources:
https://kb.froglogic.com/display/KB/Problem+-+Bringing+window+to+foreground+%28Qt%29#Problem-Bringingwindowtoforeground%28Qt%29-IMPORTANTAdditionalsettingsforWindows
https://docs.microsoft.com/en-us/previous-versions/windows/it-pro/windows-2000-server/cc957208(v=technet.10)

Maybe its worth looking into using the **win32gui** module since we already have it installed. I haven't confirmed that the **ForegroundLockTimeout** is the issue. 

Sources:
http://www.blog.pythonlibrary.org/2014/10/20/pywin32-how-to-bring-a-window-to-front/

**Comment by jnbarchan** [comment]
I have now reproduced the different behaviour in a standalone test under Windows.  It will not be **ForegroundLockTimeout**.  I will be fixing in non-platform-specific Qt code shortly.

**Comment by jnbarchan** [comment]
It seems that, under Windows at least, we need to call `preview.raise_()` at some time *after* the invoking (modal) dialog has closed itself.

Now, the modeless dialog opening code has no knowledge of where it is being called from, and I don't want to have to change the code of the calling dialog(s) to do the work there.  So I have gone for:
```
# so we do it on a timer here
# the delay is a short half-second
# we will have to see whether this always proves sufficient...
QTimer.singleShot(500, self.modelessHtmlPreviewDialogBringUpfrontAfterShowing)
```

**Comment by LaurenceLeo** [comment]
This works fine for me now. I also just used your update script to update my code. Also works great!
